File: src/client/load-test.ts
import { MusicStreamingClient } from './music-streaming.client';

/**
 * Resultado de um teste de carga
 */
export interface LoadTestResult {
  technology: string;
  operation: string;
  totalRequests: number;
  successfulRequests: number;
  failedRequests: number;
  averageTime: number;
  minTime: number;
  maxTime: number;
  medianTime: number;
  p95Time: number;
  p99Time: number;
  requestsPerSecond: number;
  errorRate: number;
}

/**
 * Tester de carga para comparar tecnologias
 */
export class LoadTester {
  private client: MusicStreamingClient;
  private results: LoadTestResult[] = [];

  constructor() {
    this.client = new MusicStreamingClient();
  }

  /**
   * Executar teste de carga para uma operação
   */
  private async runLoadTest(
    technology: string,
    operation: string,
    operationFn: () => Promise<any>,
    numberOfRequests: number = 100,
  ): Promise<LoadTestResult> {
    console.log(
      `\n⏱️  Testando ${technology.padEnd(10)} - ${operation.padEnd(30)} (${numberOfRequests} req)`,
    );

    const times: number[] = [];
    let successful = 0;
    let failed = 0;

    const startTime = performance.now();

    for (let i = 0; i < numberOfRequests; i++) {
      const reqStart = performance.now();
      try {
        await operationFn();
        successful++;
      } catch (error) {
        failed++;
      }
      times.push(performance.now() - reqStart);

      // Mostrar progresso
      if ((i + 1) % Math.ceil(numberOfRequests / 20) === 0) {
        process.stdout.write('.');
      }
    }

    const totalTime = (performance.now() - startTime) / 1000;

    // Calcular estatísticas
    times.sort((a, b) => a - b);
    const avgTime = times.reduce((a, b) => a + b, 0) / times.length;
    const medianTime = times[Math.floor(times.length / 2)];
    const p95Time = times[Math.ceil(times.length * 0.95)];
    const p99Time = times[Math.ceil(times.length * 0.99)];

    const result: LoadTestResult = {
      technology,
      operation,
      totalRequests: numberOfRequests,
      successfulRequests: successful,
      failedRequests: failed,
      averageTime: avgTime,
      minTime: Math.min(...times),
      maxTime: Math.max(...times),
      medianTime,
      p95Time,
      p99Time,
      requestsPerSecond: numberOfRequests / totalTime,
      errorRate: (failed / numberOfRequests) * 100,
    };

    console.log(' ✅');
    console.log(
      `   Tempo médio: ${result.averageTime.toFixed(2)}ms | Req/s: ${result.requestsPerSecond.toFixed(2)}`,
    );
    console.log(
      `   Min: ${result.minTime.toFixed(2)}ms | Max: ${result.maxTime.toFixed(2)}ms | P95: ${result.p95Time.toFixed(2)}ms`,
    );

    this.results.push(result);
    return result;
  }

  /**
   * Executar suite completa de testes
   */
  async runFullLoadTest(numberOfRequests: number = 100) {
    console.log('🚀 TESTE DE CARGA COMPLETO\n');
    console.log(`📊 Requisições por teste: ${numberOfRequests}\n`);
    console.log('═'.repeat(90));

    const operations = [
      {
        name: 'Listar Usuários',
        restFn: () => this.client.restListAllUsers(),
        graphqlFn: () => this.client.graphqlListAllUsers(),
        soapFn: () => this.client.soapListAllUsers(),
        grpcFn: () => this.client.grpcListAllUsers(),
      },
      {
        name: 'Listar Músicas',
        restFn: () => this.client.restListAllMusics(),
        graphqlFn: () => this.client.graphqlListAllMusics(),
        soapFn: () => this.client.soapListAllMusics(),
        grpcFn: () => this.client.grpcListAllMusics(),
      },
      {
        name: 'Playlists do Usuário (ID=1)',
        restFn: () => this.client.restListUserPlaylists(1),
        graphqlFn: () => this.client.graphqlListUserPlaylists(1),
        soapFn: () => this.client.soapListUserPlaylists(1),
        grpcFn: () => this.client.grpcListUserPlaylists(1),
      },
      {
        name: 'Músicas da Playlist (ID=1)',
        restFn: () => this.client.restListPlaylistMusics(1),
        graphqlFn: () => this.client.graphqlListPlaylistMusics(1),
        soapFn: () => this.client.soapListPlaylistMusics(1),
        grpcFn: () => this.client.grpcListPlaylistMusics(1),
      },
      {
        name: 'Playlists com Música (ID=1)',
        restFn: () => this.client.restListPlaylistsByMusic(1),
        graphqlFn: () => this.client.graphqlListPlaylistsByMusic(1),
        soapFn: () => this.client.soapListPlaylistsByMusic(1),
        grpcFn: () => this.client.grpcListPlaylistsByMusic(1),
      },
    ];

    // Executar testes
    for (const op of operations) {
      await this.runLoadTest('REST', op.name, op.restFn, numberOfRequests);
      await this.runLoadTest('GraphQL', op.name, op.graphqlFn, numberOfRequests);
      await this.runLoadTest('SOAP', op.name, op.soapFn, numberOfRequests);
      await this.runLoadTest('gRPC', op.name, op.grpcFn, numberOfRequests);
    }

    // Gerar relatório
    this.generateReport();
  }

  /**
   * Teste de estresse - aumentar carga progressivamente
   */
  async runStressTest() {
    console.log('\n💪 TESTE DE ESTRESSE (CARGA PROGRESSIVA)\n');
    console.log('═'.repeat(90));

    const cargas = [10, 50, 100, 200, 500];

    for (const carga of cargas) {
      console.log(`\n${'─'.repeat(90)}`);
      console.log(`📍 CARGA: ${carga} requisições`);
      console.log('─'.repeat(90));

      this.results = [];

      await this.runLoadTest(
        'REST',
        'Listar Usuários',
        () => this.client.restListAllUsers(),
        carga,
      );

      await this.runLoadTest(
        'GraphQL',
        'Listar Usuários',
        () => this.client.graphqlListAllUsers(),
        carga,
      );

      await this.runLoadTest(
        'SOAP',
        'Listar Usuários',
        () => this.client.soapListAllUsers(),
        carga,
      );

      await this.runLoadTest(
        'gRPC',
        'Listar Usuários',
        () => this.client.grpcListAllUsers(),
        carga,
      );

      // Mostrar resumo da carga
      this.printCargoSummary(carga);
    }
  }

  /**
   * Imprimir resumo de uma carga
   */
  private printCargoSummary(carga: number) {
    console.log(`\n📈 Resumo da carga ${carga}:`);

    const byTech = new Map<string, LoadTestResult[]>();
    this.results.forEach((r) => {
      if (!byTech.has(r.technology)) {
        byTech.set(r.technology, []);
      }
      byTech.get(r.technology)!.push(r);
    });

    byTech.forEach((results, tech) => {
      const avgTime = results.reduce((a, b) => a + b.averageTime, 0) / results.length;
      const avgRps = results.reduce((a, b) => a + b.requestsPerSecond, 0) / results.length;
      const totalErrors = results.reduce((a, b) => a + b.failedRequests, 0);

      console.log(
        `  ${tech.padEnd(10)}: ${avgTime.toFixed(2)}ms | ${avgRps.toFixed(2)} req/s | Erros: ${totalErrors}`,
      );
    });
  }

  /**
   * Gerar relatório comparativo detalhado
   */
  private generateReport() {
    console.log('\n\n' + '═'.repeat(90));
    console.log('📊 RELATÓRIO COMPARATIVO COMPLETO');
    console.log('═'.repeat(90));

    // Agrupar por tecnologia
    const byTechnology = new Map<string, LoadTestResult[]>();
    this.results.forEach((result) => {
      if (!byTechnology.has(result.technology)) {
        byTechnology.set(result.technology, []);
      }
      byTechnology.get(result.technology)!.push(result);
    });

    // Exibir relatório por tecnologia
    byTechnology.forEach((results, tech) => {
      console.log(`\n🔹 ${tech}\n`);
      console.log(
        'Operação'.padEnd(35) +
          'Tempo Médio (ms)'.padEnd(18) +
          'Req/s'.padEnd(12) +
          'Taxa Sucesso'.padEnd(15) +
          'P95 (ms)',
      );
      console.log('─'.repeat(90));

      results.forEach((result) => {
        const successRate = ((result.successfulRequests / result.totalRequests) * 100).toFixed(1);
        console.log(
          result.operation.padEnd(35) +
            result.averageTime.toFixed(2).padEnd(18) +
            result.requestsPerSecond.toFixed(2).padEnd(12) +
            `${successRate}%`.padEnd(15) +
            result.p95Time.toFixed(2),
        );
      });

      const avgTime = results.reduce((a, b) => a + b.averageTime, 0) / results.length;
      const avgRps = results.reduce((a, b) => a + b.requestsPerSecond, 0) / results.length;
      const avgP95 = results.reduce((a, b) => a + b.p95Time, 0) / results.length;

      console.log('─'.repeat(90));
      console.log(
        'MÉDIA'.padEnd(35) +
          avgTime.toFixed(2).padEnd(18) +
          avgRps.toFixed(2).padEnd(12) +
          '100.0%'.padEnd(15) +
          avgP95.toFixed(2),
      );
    });

    // Comparação entre tecnologias
    console.log('\n\n' + '─'.repeat(90));
    console.log('🏆 RANKING DE TECNOLOGIAS');
    console.log('─'.repeat(90));

    const techComparison = new Map<string, { totalTime: number; totalRps: number; count: number }>();

    this.results.forEach((result) => {
      if (!techComparison.has(result.technology)) {
        techComparison.set(result.technology, {
          totalTime: 0,
          totalRps: 0,
          count: 0,
        });
      }
      const current = techComparison.get(result.technology)!;
      current.totalTime += result.averageTime;
      current.totalRps += result.requestsPerSecond;
      current.count += 1;
    });

    const sorted = Array.from(techComparison.entries())
      .map(([tech, data]) => ({
        tech,
        avgTime: data.totalTime / data.count,
        avgRps: data.totalRps / data.count,
      }))
      .sort((a, b) => a.avgTime - b.avgTime);

    console.log('Posição'.padEnd(10) + 'Tecnologia'.padEnd(15) + 'Tempo Médio (ms)'.padEnd(20) + 'Req/s Médio');
    console.log('─'.repeat(90));

    sorted.forEach((item, index) => {
      const medals = ['🥇', '🥈', '🥉'];
      const medal = medals[index] || '  ';
      const position = `${index + 1}°`;
      console.log(
        position.padEnd(10) +
          item.tech.padEnd(15) +
          item.avgTime.toFixed(2).padEnd(20) +
          item.avgRps.toFixed(2),
      );
      console.log(`${medal} ${item.tech} é ${index === 0 ? 'mais rápido' : index === 1 ? 'intermediário' : 'mais lento'}\n`);
    });

    // Estatísticas gerais
    console.log('─'.repeat(90));
    console.log('\n📌 ESTATÍSTICAS GERAIS\n');
    console.log(`Total de requisições: ${this.results.reduce((a, b) => a + b.totalRequests, 0)}`);
    console.log(`Total de sucessos: ${this.results.reduce((a, b) => a + b.successfulRequests, 0)}`);
    console.log(`Total de falhas: ${this.results.reduce((a, b) => a + b.failedRequests, 0)}`);

    const totalTime = this.results.reduce((a, b) => a + b.averageTime, 0) / this.results.length;
    const totalRps = this.results.reduce((a, b) => a + b.requestsPerSecond, 0) / this.results.length;

    console.log(`Tempo médio geral: ${totalTime.toFixed(2)}ms`);
    console.log(`Req/s médio geral: ${totalRps.toFixed(2)}`);

    console.log('\n' + '═'.repeat(90));
    console.log('✅ Relatório completo gerado!\n');
  }
}

// Executar quando chamado diretamente
async function main() {
  const tester = new LoadTester();

  const args = process.argv.slice(2);
  const stressMode = args.includes('--stress');
  const requestCount = parseInt(args.find((a) => a.startsWith('--requests='))?.split('=')[1] || '100');

  if (stressMode) {
    await tester.runStressTest();
  } else {
    await tester.runFullLoadTest(requestCount);
  }
}

if (require.main === module) {
  main().catch(console.error);
}

export default LoadTester;

File: client_python/load_test.py
"""
Tester de carga para comparar tecnologias de música streaming
"""

import asyncio
import time
from typing import List, Dict, Callable, Any
from dataclasses import dataclass
from statistics import mean, median, stdev
import sys
from music_streaming_client import MusicStreamingClient, ClientConfig


@dataclass
class LoadTestResult:
    """Resultado de um teste de carga"""
    technology: str
    operation: str
    total_requests: int
    successful_requests: int
    failed_requests: int
    average_time: float
    min_time: float
    max_time: float
    median_time: float
    p95_time: float
    p99_time: float
    requests_per_second: float
    error_rate: float


class LoadTester:
    """Tester de carga para comparar tecnologias"""

    def __init__(self):
        """Inicializar tester"""
        self.client = MusicStreamingClient()
        self.results: List[LoadTestResult] = []

    async def _run_load_test(
        self,
        technology: str,
        operation: str,
        operation_fn: Callable,
        number_of_requests: int = 100,
    ) -> LoadTestResult:
        """Executar teste de carga para uma operação"""
        print(f"\n⏱️  Testando {technology:10} - {operation:30} ({number_of_requests} req)")

        times: List[float] = []
        successful = 0
        failed = 0

        start_time = time.time()

        for i in range(number_of_requests):
            req_start = time.time()
            try:
                await operation_fn()
                successful += 1
            except Exception as error:
                failed += 1
            times.append((time.time() - req_start) * 1000)  # Converter para ms

            # Mostrar progresso
            if (i + 1) % max(1, number_of_requests // 20) == 0:
                sys.stdout.write(".")
                sys.stdout.flush()

        total_time = time.time() - start_time

        # Calcular estatísticas
        times.sort()
        avg_time = mean(times)
        min_time = min(times)
        max_time = max(times)
        median_time = median(times)
        p95_time = times[int(len(times) * 0.95)]
        p99_time = times[int(len(times) * 0.99)]

        result = LoadTestResult(
            technology=technology,
            operation=operation,
            total_requests=number_of_requests,
            successful_requests=successful,
            failed_requests=failed,
            average_time=avg_time,
            min_time=min_time,
            max_time=max_time,
            median_time=median_time,
            p95_time=p95_time,
            p99_time=p99_time,
            requests_per_second=number_of_requests / total_time,
            error_rate=(failed / number_of_requests) * 100,
        )

        print(" ✅")
        print(
            f"   Tempo médio: {result.average_time:.2f}ms | Req/s: {result.requests_per_second:.2f}"
        )
        print(
            f"   Min: {result.min_time:.2f}ms | Max: {result.max_time:.2f}ms | P95: {result.p95_time:.2f}ms"
        )

        self.results.append(result)
        return result

    async def run_full_load_test(self, requests_per_operation: int = 100) -> None:
        """Executar teste completo com todas as tecnologias"""
        print("\n" + "=" * 80)
        print("🚀 TESTE DE CARGA COMPLETO - COMPARAÇÃO DE TECNOLOGIAS")
        print("=" * 80)

        # Operações a testar
        operations = [
            ("Listar Usuários", lambda: self.client.rest_list_all_users(), lambda: self.client.graphql_list_all_users(), lambda: self.client.soap_list_all_users(), lambda: self.client.grpc_list_all_users()),
            ("Listar Músicas", lambda: self.client.rest_list_all_musics(), lambda: self.client.graphql_list_all_musics(), lambda: self.client.soap_list_all_musics(), lambda: self.client.grpc_list_all_musics()),
            ("Playlists do Usuário (ID=1)", lambda: self.client.rest_list_user_playlists(1), lambda: self.client.graphql_list_user_playlists(1), lambda: self.client.soap_list_user_playlists(1), lambda: self.client.grpc_list_user_playlists(1)),
            ("Músicas da Playlist (ID=1)", lambda: self.client.rest_list_playlist_musics(1), lambda: self.client.graphql_list_playlist_musics(1), lambda: self.client.soap_list_playlist_musics(1), lambda: self.client.grpc_list_playlist_musics(1)),
            ("Playlists com Música (ID=1)", lambda: self.client.rest_list_playlists_by_music(1), lambda: self.client.graphql_list_playlists_by_music(1), lambda: self.client.soap_list_playlists_by_music(1), lambda: self.client.grpc_list_playlists_by_music(1)),
        ]

        # Testar cada tecnologia
        for op_name, rest_fn, graphql_fn, soap_fn, grpc_fn in operations:
            await self._run_load_test("REST", op_name, rest_fn, requests_per_operation)
            await self._run_load_test("GraphQL", op_name, graphql_fn, requests_per_operation)
            await self._run_load_test("SOAP", op_name, soap_fn, requests_per_operation)
            await self._run_load_test("gRPC", op_name, grpc_fn, requests_per_operation)

        self._print_summary()

    def _print_summary(self) -> None:
        """Imprimir sumário dos testes"""
        print("\n" + "=" * 80)
        print("📊 RELATÓRIO COMPARATIVO COMPLETO")
        print("=" * 80)

        # Agrupar resultados por tecnologia
        tech_results = {}
        for result in self.results:
            if result.technology not in tech_results:
                tech_results[result.technology] = []
            tech_results[result.technology].append(result)

        # Imprimir relatório para cada tecnologia
        for technology in ["REST", "GraphQL", "SOAP", "gRPC"]:
            if technology in tech_results:
                self._print_technology_report(technology, tech_results[technology])

        # Imprimir ranking
        self._print_ranking()

    def _print_technology_report(self, technology: str, results: List[LoadTestResult]) -> None:
        """Imprimir relatório de uma tecnologia"""
        print(f"\n🌐 {technology}\n")
        print(f"{'Operação':<32} {'Tempo Médio (ms)':<18} {'Req/s':<15} {'Taxa Sucesso':<15} {'P95 (ms)':<12}")
        print("-" * 92)

        avg_time = 0
        avg_req_per_sec = 0

        for result in results:
            success_rate = ((result.total_requests - result.failed_requests) / result.total_requests) * 100
            print(
                f"{result.operation:<32} {result.average_time:<18.2f} {result.requests_per_second:<15.2f} {success_rate:<14.1f}% {result.p95_time:<12.2f}"
            )
            avg_time += result.average_time
            avg_req_per_sec += result.requests_per_second

        print("-" * 92)
        avg_time /= len(results)
        avg_req_per_sec /= len(results)
        print(f"{'MÉDIA':<32} {avg_time:<18.2f} {avg_req_per_sec:<15.2f} {'100.0%':<14} {0:<12.2f}")

    def _print_ranking(self) -> None:
        """Imprimir ranking das tecnologias"""
        print("\n" + "=" * 80)
        print("🏆 RANKING DE TECNOLOGIAS")
        print("=" * 80)

        # Calcular média de tempo para cada tecnologia
        tech_avg_time = {}
        tech_avg_req_per_sec = {}

        for technology in ["REST", "GraphQL", "SOAP", "gRPC"]:
            times = [r.average_time for r in self.results if r.technology == technology]
            req_per_sec = [r.requests_per_second for r in self.results if r.technology == technology]
            
            if times:
                tech_avg_time[technology] = mean(times)
                tech_avg_req_per_sec[technology] = mean(req_per_sec)

        # Ordenar por tempo médio
        ranked = sorted(tech_avg_time.items(), key=lambda x: x[1])

        print(f"\n{'Posição':<12} {'Tecnologia':<15} {'Tempo Médio (ms)':<20} {'Req/s Médio':<15}")
        print("-" * 62)

        for i, (technology, avg_time) in enumerate(ranked, 1):
            print(
                f"{i}º{'':<10} {technology:<15} {avg_time:<20.2f} {tech_avg_req_per_sec[technology]:<15.2f}"
            )

            if i == 1:
                print(f"{'':12}gRPC é mais rápido")
            elif i == 2:
                print(f"{'':12}tecnologia intermediária")
            elif i == len(ranked):
                print(f"{'':12}tecnologia mais lenta")

        # Estatísticas gerais
        total_requests = sum(r.total_requests for r in self.results)
        total_successful = sum(r.successful_requests for r in self.results)
        avg_time_general = mean([r.average_time for r in self.results])
        avg_req_per_sec_general = mean([r.requests_per_second for r in self.results])

        print("\n" + "-" * 80)
        print("📊 ESTATÍSTICAS GERAIS\n")
        print(f"Total de requisições: {total_requests}")
        print(f"Total de sucessos: {total_successful}")
        print(f"Total de falhas: {total_requests - total_successful}")
        print(f"Tempo médio geral: {avg_time_general:.2f}ms")
        print(f"Req/s médio geral: {avg_req_per_sec_general:.2f}")

        print("\n" + "=" * 80)
        print("✅ Relatório completo gerado!")
        print("=" * 80)


async def main():
    """Função principal"""
    tester = LoadTester()
    
    # Executar teste de carga completo
    await tester.run_full_load_test(requests_per_operation=100)
    
    # Gerar gráficos automaticamente
    try:
        from generate_charts import LoadTestChartGenerator
        print("\n" + "="*80)
        print("📊 Gerando gráficos dos resultados...")
        print("="*80)
        generator = LoadTestChartGenerator()
        generator.generate_all()
        print("\n✅ Gráficos salvos em: ./charts/")
    except ImportError:
        print("\n⚠️  matplotlib não instalado para gerar gráficos")
        print("   Execute: pip install matplotlib")
    except Exception as e:
        print(f"\n⚠️  Erro ao gerar gráficos: {e}")


if __name__ == "__main__":
    asyncio.run(main())
